### PR TITLE
refactor: Port factory reset into new architecture

### PR DESCRIPTION
**Pull request by kartikCypherock**
Implements https://app.clickup.com/t/9002019994/PRF-4850

### CODE REVIEW

File: src/settings/factory_reset.c [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+/**
+ * @brief Get the uncommon wallet list between the X1 vault and two X1 cards
+ *
+ * @param wallets_in_vault Pointer to buffer filled with wallet list in X1 vault
+ * @param wallets_in_cards Pointer to buffer filled with common wallet list in
+ * two cards
+ * @param uncommon_wallets Pointer to buffer which will be filled with the
+ * wallets present in X1 vault but not in the two cards
+ */
+static void get_uncommon_wallet_list(wallet_list_t *wallets_in_vault,
```
Comment (ujjwal-cyph) [inline src/settings/factory_reset.c:125]: please see if we can reduce duplicity in logic

File: src/settings/factory_reset.c [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+  if (!core_scroll_page(NULL, ui_text_confirm_factory_reset, NULL)) {
+    return;
+  }
+
+  // If there is atleast one valid wallet on the device, ensure that it is
+  // present in atleast 2 X1 cards
+  if (0 < valid_wallets) {
+    if (!get_wallet_list_from_two_cards(&wallets_in_cards)) {
+      return;
```
Comment (ujjwal-cyph) [inline src/settings/factory_reset.c:265]: let's ensure screen flicker/stuck behavior does not appear

File: src/settings/factory_reset.c (hunk 1 above)
Comment (kartikCypherock) [reply src/settings/factory_reset.c]: I have tried by splitting the comparison logic and logic whether to add the wallet to the list or not as part of a callback.

File: src/settings/factory_reset.c (hunk 2 above)
Comment (kartikCypherock) [reply src/settings/factory_reset.c:252]: Added

File: src/settings/factory_reset.c [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+      char msg[64];
+      const char *msg_list[3] = {
+          msg, ui_text_reset_exit[0], ui_text_reset_exit[1]};
+
+      snprintf(msg,
+               sizeof(msg),
+               UI_TEXT_FACTORY_RESET_ERROR,
+               uncommon_wallets.wallet[0].name);
+
+      multi_instruction_init(msg_list, 3, DELAY_LONG_STRING, true);
```
Comment (amanCypherock) [inline src/settings/factory_reset.c:273]: Are we still keeping the multi-instruction component.

Comment (kartikCypherock) [reply src/settings/factory_reset.c:273]: As of now, yes, Since there is not discussion on correct screen

File: src/settings/factory_reset.c [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+      // Display first wallet not present in cards - just to keep it
+      // consistent with current UX
+      char msg[64];
+      const char *msg_list[3] = {
+          msg, ui_text_reset_exit[0], ui_text_reset_exit[1]};
+
+      snprintf(msg,
+               sizeof(msg),
+               UI_TEXT_FACTORY_RESET_ERROR,
+               uncommon_wallets.wallet[0].name);
```
Comment (amanCypherock) [inline src/settings/factory_reset.c:271]: Consider updating the message to indicate that multiple wallets are not present on either 1 or more than 1 card.
Let's change this in the draw.io designs first and take this change up as a separate improvement.

Comment (amanCypherock) [reply src/settings/factory_reset.c:271]: Should add a todo just to make sure we don't miss this for too long.

Comment (kartikCypherock) [reply src/settings/factory_reset.c:274]: Added

**Review by amanCypherock — APPROVED** [review]

**Review by ujjwal-cyph — APPROVED** [review]
LGTM

**Pull request merged**